exonum/src/api/manager.rs

```diff
@@ -100,35 +102,69 @@ impl ApiManager {
 impl Actor for ApiManager {
     type Context = Context<Self>;
 
-    fn started(&mut self, _ctx: &mut Self::Context) {
+    fn started(&mut self, ctx: &mut Self::Context) {
         self.start_api_servers();
+        let endpoints_rx = self
+            .endpoints_rx
+            .take()
+            .expect("`Actor::started()` called twice for `ApiManager`");
+        Self::add_stream(endpoints_rx, ctx);
     }
 }
 
-impl Message for RestartServer {
+#[derive(Debug)]
+struct StartServer {
+    config: ApiRuntimeConfig,
+    attempt: usize,
+}
+
+impl Message for StartServer {
     type Result = ();
 }
 
-impl Handler<RestartServer> for ApiManager {
+impl Handler<StartServer> for ApiManager {
     type Result = ();
 
-    fn handle(&mut self, _msg: RestartServer, ctx: &mut Context<Self>) -> Self::Result {
-        info!("Server restart requested");
-        self.initiate_restart(ctx.address());
+    fn handle(&mut self, mut msg: StartServer, ctx: &mut Context<Self>) -> Self::Result {
+        info!(
+            "Handling server start: {:?} (attempt #{})",
+            msg.config,
+            msg.attempt + 1
+        );
+        let addr = match self.start_server(msg.config.clone()) {
+            Ok(addr) => addr,
+            Err(e) => {
+                warn!("Error handling service start {:?}: {}", msg.config, e);
+                if msg.attempt == Self::RETRY_ATTEMPTS {
+                    error!("Cannot spawn server with config {:?}", msg.config);
+                    ctx.terminate();
+                } else {
+                    msg.attempt += 1;
+                    ctx.notify_later(msg, Self::RETRY_INTERVAL);
+                }
+                return;
+            }
+        };
+        self.api_runtime_addresses.insert(addr, msg.config);
     }
 }
 
-impl Message for StartServer {
-    type Result = ();
+/// Updates user-provided endpoints, restarting all HTTP service managed by the addressed
+/// `ApiManager`.
+#[derive(Debug, Clone)]
+pub struct UpdateEndpoints {
```

Comment:
In the case there are several producers of endpoints (e.g., several runtimes), this struct needs to contain the ID of the changed part in order not to erase parts corresponding to other producers. Seems easy to implement, but I'm not sure whether this is necessary right now (we don't have other producers besides `RustRuntime`).